provides.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2021 Canonical Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from charms import reactive
+import charmhelpers.contrib.network.ip as ch_net_ip
+
+
+class BindClientProvides(reactive.Endpoint):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.ingress_address = ch_net_ip.get_relation_ip(self.endpoint_name)
+
+    def relation_ids(self):
+        return [relation.relation_id for relation in self.relations]
+
+    def set_ingress_address(self):
+        for relation in self.relations:
+            relation.to_publish_raw["ingress-address"] = self.ingress_address
+            relation.to_publish_raw["private-address"] = self.ingress_address
+
+    def available(self):
+        return reactive.is_flag_set(
+            self.expand_name("{endpoint_name}.available")
+        )
+
+    @reactive.when("endpoint.{endpoint_name}.joined")
+    def joined(self):
+        if not self.available():
+            reactive.set_flag(self.expand_name("{endpoint_name}.available"))
+            self.set_ingress_address()
+
+    def remove(self):
+        if self.available():
+            reactive.clear_flag(self.expand_name("{endpoint_name}.available"))
+
+    @reactive.when("endpoint.{endpoint_name}.broken")
+    def broken(self):
+        self.remove()
+
+    @reactive.when("endpoint.{endpoint_name}.departed")
+    def departed(self):
+        self.remove()
```

Comment:
```suggestion
    @reactive.when_any("endpoint.{endpoint_name}.broken", "endpoint.{endpoint_name}.departed")
    def broken(self):
        if self.available():
            reactive.clear_flag(self.expand_name("{endpoint_name}.available"))
```

Reply:
@auria and @r-f-g The problem that happens when it uses `when_any` is that the handler never gets the `self` argument as it says in the [charm-reactive](https://github.com/juju-solutions/charms.reactive/blob/master/charms/reactive/decorators.py#L134-L143). I've tried also to make as a classmethod, but it doesn't work also as it shows in this [logs](https://pastebin.canonical.com/p/ssfbF7zY25/)

This part of the [documentation](https://charmsreactive.readthedocs.io/en/latest/charms.reactive.decorators.html#charms-reactive-decorators) also confirms this behavior of not passing arguments.

I think we should keep the way as it is implemented with one flag for each endpoint handler.

Reply:
@r-f-g  I discovered that the problem was that the framework is not capable of recognize the endpoint if you use `class BindClientProvides(reactive.Endpoint)` instead of `class BindClientProvides(Endpoint)`. I think this is kind of a bug, but in anyway, I've changed and now it works.

Thanks for checking this.